### PR TITLE
Narrow cas-add inline size error to MCP handler only

### PR DESCRIPTION
**Pull request by sergey-shandar**
## Summary

This PR narrows the scope of the `cas-add-inline-size-error` issue to focus exclusively on the MCP `cas_add` handler. The broader codec groundwork (bounded `tryListToVec` / `tryU8ListToVec`, `tryUtf8`, nullable propagation) is already in place; this issue now tracks only the handler-level changes needed to reject oversized inline content as a clean decoding error on every JS engine.

## Changes

The issue description has been significantly condensed to reflect the narrowed scope:

- **Removed** the extensive multi-step proposal for propagating `Nullable<Vec>` through the entire codebase (bit vectors, ASN.1 encoders, file I/O, HTTP boundaries, etc.). That work is either already complete or tracked separately.

- **Removed** detailed implementation guidance for `unpackListToVec`, `fs/base64` `decode`, `fs/text` `utf8`, and the `unwrap` helper — these are already implemented or out of scope.

- **Kept** the core problem statement: `cas_add` does not yet turn over-cap inline payloads into a clean error on every engine (on `bun`, the throwing `utf8` crashes the MCP server; on `node`/`deno`, `base64Decode` returns `null` but the error message is misleading for the size case).

- **Simplified the proposal** to three concrete handler-level tasks:
  1. Switch `text` branch from total `utf8` to non-throwing `tryUtf8`
  2. Map `null` from both `tryUtf8` and `base64Decode` to a single generic decoding error (static message, no content echo, no cause branching)
  3. Optional up-front length check as an optimization

- **Clarified scope boundaries**: two remaining base64 edge cases (`encode-padding-overflow.md`, `decode-rejects-max-size-input.md`) are tracked separately and out of scope here.

- **Updated task list** to reflect handler-only changes and proof tests for the `cas_add` boundary.

## Implementation Details

The fix ensures that:
- An over-`maxLength` `text` payload yields `null` (not a thrown `bigint`) via `tryUtf8`, preventing crashes on `bun`
- Both `null` cases (malformed or oversized) are treated identically as a decoding error
- The error message restates the documented 128 KiB inline limit and points callers to `type: 'url'` for large content
- No `unwrap` is used in the handler — the over-cap signal ends only as the `isError` result

https://claude.ai/code/session_01UzbMLSjzmAYfwvVucakGEe

### CODE REVIEW

**Comment by sergey-shandar** [comment]
@codex review

File: fs/cas/todo/cas-add-inline-size-error.md [hunk 1]
```diff
@@ -1,307 +1,110 @@
...
+- `fs/text/module.f.ts` has `tryUtf8` (`Nullable<Utf8>`, `null` only past
+  `maxLength`); `utf8` is the total `mapUnwrap` wrapper.
+- `fs/base64/module.f.ts` `decode` returns `null` for over-`maxLength` input
+  (its `stringToVec` core is now `Nullable`) as well as for malformed input.
 
-The defect is only that the cap is **not enforced as a clean error**. The conversion
-that builds the `Vec` runs *before* the `cas_add` handler can react:
+The two remaining base64 edge cases are tracked separately and are **out of scope
+here**: `fs/base64/todo/encode-padding-overflow.md` and
+`fs/base64/todo/decode-rejects-max-size-input.md`.
```
[the remark this answers is not in the document]
Comment (sergey-shandar) [reply fs/cas/todo/cas-add-inline-size-error.md]: Fixed in 07e0704. `decode-rejects-max-size-input` is no longer "out of scope" — it's now listed under **Blocked by** as a prerequisite for the exact-`maxLengthBytes` base64 *success* case, and the proof-test task is split: the `text` boundary and both one-byte-over cases work with the handler change alone, while the base64 exact-size success is gated on the decode fix (until then the boundary input asserts the generic `decoding error`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: fs/cas/todo/cas-add-inline-size-error.md [hunk 2]
```diff
@@ -1,307 +1,110 @@
...
-todo's goal: a detectable over-cap signal everywhere it can arise, resolved to a typed
-error at the boundary. It also makes the design **review-stable**: a newly-found unbounded
-caller is already `Nullable`, so the type-checker forces it to be handled instead of
-silently crashing.
+3. **Additional up-front check (optional, cheap).** Before decoding, a UTF-16
+   `content.length` is a fast lower bound on the encoded byte length; an explicit
+   guard can short-circuit the obvious over-cap case with the same `decoding
+   error` before building any `Vec`. The authoritative check stays the `null`
+   from `tryUtf8` / `base64Decode` (UTF-8 expansion and base64's 3/4 ratio mean
+   string length alone is not exact), so this is only an optimisation, not a
```
[the remark this answers is not in the document]
Comment (sergey-shandar) [reply fs/cas/todo/cas-add-inline-size-error.md]: Fixed in 07e0704. The optional `content.length` short-circuit is now scoped to `text` only (where UTF-16 length is a valid lower bound on UTF-8 byte length). For `base64` the doc now explicitly says not to test `content.length` against `maxLengthBytes` (≈ 4/3 of decoded bytes) and to rely on `base64Decode`'s `null`, or derive the exact decoded length from body length + padding.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sergey-shandar** [comment]
@codex review

File: fs/cas/mcp/todo/cas-add-inline-size-error.md [hunk 3]
```diff
@@ -1,307 +1,136 @@
...
 **Status:** open
-**Blocked by:** —
+**Blocked by:** `fs/base64/todo/decode-rejects-max-size-input.md` — the exact-`maxLengthBytes`
+base64 *success* case only; the handler change and the over-cap → `decoding error`
+behaviour do not depend on it.
+
+### Scope
+
+This issue is now narrowed to the **MCP `cas_add` handler**
+(`fs/cas/mcp/module.f.ts`). The bit-vector / codec groundwork it used to describe
```
[the remark this answers is not in the document]
Comment (sergey-shandar) [reply fs/cas/mcp/todo/cas-add-inline-size-error.md:12]: Done in 5a9f7fb — `git mv`d the todo from `fs/cas/todo/` to `fs/cas/mcp/todo/cas-add-inline-size-error.md`, next to the `cas_add` handler it now describes.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sergey-shandar** [comment]
@codex review

File: fs/cas/mcp/todo/cas-add-inline-size-error.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+      - Add a `base64OfA(n)` helper spelling out base64 of `n` ASCII `'a'` bytes
+        for every `n % 3` residue, so the boundary sample never drives the encoder
+        past `maxLength`.
+- [ ] Confirm the inline-limit wording in `fs/cas/mcp/README.md` and the `cas_add`
+      tool description still matches the implemented behaviour.
+
+### Related
+
+- `fs/cas/mcp/module.f.ts` — the `cas_add` handler (and the `cas_get`
+  `content: true` oversized-blob guard, already implemented, that it mirrors).
```
[the remark this answers is not in the document]
Comment (sergey-shandar) [reply fs/cas/mcp/todo/cas-add-inline-size-error.md]: Good catch — filed as its own issue in 4a0ab60: `fs/cas/mcp/todo/cas-get-response-overflow.md`. It covers both halves: tightening the `cas_get` guard to check the serialized response (base64 inflation + JSON envelope), not just raw blob length, and switching `fs/mcp/stdio/module.f.ts` `writeResponse` from the throwing `utf8` to `tryUtf8` so an oversized line can't crash the transport. Cross-referenced from `cas-add-inline-size-error.md`'s Related section.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sergey-shandar** [comment]
@codex review

**Comment by sergey-shandar** [comment]
@codex review

File: fs/cas/mcp/todo/cas-get-response-overflow.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+throw across an MCP boundary. Both narrow the same class of bug `cas_add`'s
+inline-size handling already addresses on the input side; this is the
+output-side counterpart.
+
+### Tasks
+
+- [ ] `fs/cas/mcp/module.f.ts` `cas_get`: replace the raw `length >
+      maxLengthBytes` check (the `content: true` branch) with a check against
+      the serialized response size for the chosen `type` (base64 inflation +
+      JSON envelope for `base64`; UTF-8 bytes + envelope for `text`).
```
[the remark this answers is not in the document]
Comment (sergey-shandar) [reply fs/cas/mcp/todo/cas-get-response-overflow.md]: Good catch — confirmed: `okResult(toJson(...))` puts the inner CAS JSON string into `ToolsCallResult.content[0].text`, and `writeResponse` serializes the whole JSON-RPC envelope again, so quote/backslash/control characters in `type:'text'` content get escaped twice (e.g. `"` → `\"` → `\\\"`, ~4×). Fixed in 764c762: reworked the proposal so the transport-level `tryUtf8` fix (task 1) is the authoritative, exact backstop — it must hold regardless of estimation error — and the `cas_get`-level guard (task 2) is now explicitly a conservative, never-unsafe early reject using a worst-case 4× multiplier for `text` (base64's alphabet has no JSON-special characters, so that bound stays exact). Added a proof-test bullet for a `"`/`\`/control-character text blob whose raw size is fine but whose double-escaped line would overflow, to exercise the `tryUtf8` fallback path specifically.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: fs/cas/mcp/todo/cas-add-inline-size-error.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
+## cas-add-inline-size-error. Reject oversized `cas_add` inline content as a `decoding error`
```
[the remark this answers is not in the document]
Comment (sergey-shandar) [reply fs/cas/mcp/todo/cas-add-inline-size-error.md:1]: Fixed in 764c762 — `fs/cas/mcp/README.md`'s link now points at `todo/cas-add-inline-size-error.md` (same-directory `todo/`, since the issue moved to `fs/cas/mcp/todo/`) instead of the now-dead `../todo/`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sergey-shandar** [comment]
@codex review

File: fs/cas/mcp/todo/cas-get-response-overflow.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+### Tasks
+
+- [ ] `fs/mcp/stdio/module.f.ts` `writeResponse`: switch from `utf8` to
+      `tryUtf8`; on `null`, write a typed fallback error response instead of
+      throwing. (Authoritative fix — required regardless of task below.)
+- [ ] `fs/cas/mcp/module.f.ts` `cas_get`: replace the raw `length >
+      maxLengthBytes` check (the `content: true` branch) with a conservative
+      check against the *escaped* response size: exact `ceil(length/3)*4` +
+      envelope for `base64`; `length × 4` (worst-case escaping) + envelope for
+      `text`.
```
[the remark this answers is not in the document]
Comment (sergey-shandar) [reply fs/cas/mcp/todo/cas-get-response-overflow.md]: Right — a raw NUL becomes the 6-character `\u00XX` escape on the inner pass, and that escape's own backslash gets re-escaped on the outer pass, so it's 7 bytes for 1 raw byte, not 4. Fixed in 5a453f5: the proposal and task now use a 7× multiplier for `type:'text'`, explicitly called out as the worst case (control characters), not quote/backslash (which only hit 4×).

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sergey-shandar** [comment]
@codex review
